### PR TITLE
KAFKA-5562: execute state dir cleanup on single thread

### PR DESCRIPTION
**Pull request by dguy**
Use a single `StateDirectory` per streams instance.
Use threadId to determine which thread owns the lock.
Only allow the owning thread to unlock.
Execute cleanup on a scheduled thread in `KafkaStreams`

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang @mjsax @enothereska @bbejeck

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6014/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5999/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by dguy** [comment]
System test results: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2017-07-10--001.1499723012--dguy--kafka-5562--3ef084d/report.html

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 1]
```diff
@@ -485,7 +497,7 @@ public KafkaStreams(final TopologyBuilder builder,
                                           metrics,
                                           time,
                                           streamsMetadataState,
-                                          cacheSizeBytes);
+                                          cacheSizeBytes, stateDirectory);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Nit: move to new line

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 2]
```diff
@@ -204,12 +221,11 @@ void unlock(final TaskId taskId) throws IOException {
      * @param cleanupDelayMs only remove directories if they haven't been modified for at least
      *                       this amount of time (milliseconds)
      */
-    public void cleanRemovedTasks(final long cleanupDelayMs) {
+    public synchronized void cleanRemovedTasks(final long cleanupDelayMs) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:220]: As we schedule this in an own thread and remove from main poll-loop, I think we don't need parameter `cleanupDelayMs` anymore.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 3]
```diff
@@ -424,6 +425,57 @@ public void testToString() {
...
+        final CountDownLatch latch = new CountDownLatch(1);
+        streams.setStateListener(new KafkaStreams.StateListener() {
+            @Override
+            public void onChange(final KafkaStreams.State newState, final KafkaStreams.State oldState) {
+                if (newState == KafkaStreams.State.RUNNING && oldState == KafkaStreams.State.REBALANCING) {
+                    latch.countDown();
+                }
+            }
+        });
+        final String appDir = stateDir + "/" + appId;
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: Use `File.separator` instead of `"/"`

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 4]
```diff
@@ -424,6 +425,57 @@ public void testToString() {
...
+        final File oldTaskDir = new File(appDir, "10_1");
+        assertTrue(oldTaskDir.mkdirs());
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: Can you move this after `latch.await()` -- test logic is "create - wait for cleanup" and this should be grouped together IMHO.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java [hunk 5]
```diff
@@ -266,4 +268,51 @@ public void shouldUnlockGlobalStateDirectory() throws Exception {
...
+    @Test
+    public void shouldNotLockStateDirLockedByAnotherThread() throws IOException, InterruptedException {
+        final TaskId taskId = new TaskId(0, 0);
+        final Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    directory.lock(taskId, 1);
+                } catch (IOException e) {
+                    //
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java]: Shouldn't we re-throw a `RuntimeException` to make the test fail?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java [hunk 6]
```diff
@@ -266,4 +268,51 @@ public void shouldUnlockGlobalStateDirectory() throws Exception {
...
 
+    @Test
+    public void shouldNotLockStateDirLockedByAnotherThread() throws IOException, InterruptedException {
+        final TaskId taskId = new TaskId(0, 0);
+        final Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    directory.lock(taskId, 1);
+                } catch (IOException e) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java]: Nit: `final`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java [hunk 7]
```diff
@@ -266,4 +268,51 @@ public void shouldUnlockGlobalStateDirectory() throws Exception {
...
+        final Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    directory.lock(taskId, 1);
+                    lockLatch.countDown();
+                    unlockLatch.await();
+                    directory.unlock(taskId);
+                } catch (Exception e) {
+                    //
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java]: as above

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java [hunk 8]
```diff
@@ -266,4 +268,51 @@ public void shouldUnlockGlobalStateDirectory() throws Exception {
         }
     }
 
+    @Test
+    public void shouldNotLockStateDirLockedByAnotherThread() throws IOException, InterruptedException {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:274]: Don't understand this test -- there is only one thread...

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 9]
```diff
@@ -442,16 +444,16 @@ public void testPartitionAssignmentChangeForMultipleGroups() throws Exception {
...
+                builder,
+                config,
+                clientSupplier,
+                applicationId,
+                clientId,
+                processId,
+                metrics,
+                Time.SYSTEM,
+                new StreamsMetadataState(builder, StreamsMetadataState.UNKNOWN_HOST),
+                0, stateDirectory);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Nit: move to new line. Why did you change indention?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java (hunk 2 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:220]: I think we should leave it. We should cleanup old state dirs, but leave recently touched dirs as they would be useful in case of failure

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java (hunk 4 above)
Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: I think it is correct as is. The latch is there to ensure that the stream is in the running state BEFORE verifying that the cleanup has happened

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java (hunk 8 above)
Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:274]: Hmmm, there is the main thread and the thread created in the test

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java (hunk 5 above)
Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java]: test won't fail as it is not happening on the main thread. Doesn't matter

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java (hunk 7 above)
Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java]: as above

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6112/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6097/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by enothereska** [comment]
Interesting, thanks. LGTM

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java (hunk 5 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java]: Maybe we should put the exception into a class member than and check in main thread?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java (hunk 8 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:274]: Ah... Now I get it... (is it just me, or is this test logic not obvious?)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java (hunk 7 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java]: as above 😂

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java (hunk 5 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java]: Or does `IOException` and "didn't get lock" correlate and the test would fail in L286? For this case, we should at least call `printStackTrace()` and `flush` to make debugging possible.

Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java]: The test would fail as the lock was not taken. So the assertion would fail.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java (hunk 8 above)
Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:274]: I think it is just you ;-)
It is pretty obvious to me

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6144/
Test PASSed (JDK 7 and Scala 2.11).

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java (hunk 8 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:274]: You don't count for this case -- you wrote it. ;P \cc @enothereska @bbejeck @guozhangwang

**Review by bbejeck — APPROVED** [review]
LGTM

**Comment by dguy** [comment]
@guozhangwang

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 10]
```diff
@@ -127,6 +129,14 @@
     private static final int DEFAULT_CLOSE_TIMEOUT = 0;
     private GlobalStreamThread globalStreamThread;
 
+    private final ScheduledExecutorService stateDirCleaner = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
+        @Override
+        public Thread newThread(final Runnable r) {
+            final Thread thread = new Thread(r, "kafka-streams-cleanup-thread");
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: For stream threads the thread name is as `clientId + "-StreamThread-" + STREAM_THREAD_ID_SEQUENCE`.

Should we make this thread name as `clientId + "-CleanupThread"` to be consistent?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: I know that we did not follow this consistent naming for the shutdown thread but I thought since that is a transient thread so did not want to be picky at that time; but this thread is as long living as other stream threads..

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 11]
```diff
@@ -463,12 +474,13 @@ public KafkaStreams(final TopologyBuilder builder,
         final long cacheSizeBytes = Math.max(0, config.getLong(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG) /
                 (config.getInt(StreamsConfig.NUM_STREAM_THREADS_CONFIG) + (globalTaskTopology == null ? 0 : 1)));
 
+        stateDirectory = new StateDirectory(applicationId, "main", config.getString(StreamsConfig.STATE_DIR_CONFIG), time);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Hmm.. Seems this second parameter is not useful any more since it is only for log4j prefix, while now all threads will have the same prefix.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 12]
```diff
@@ -45,21 +45,29 @@
...
+    private static class LockAndOwner {
+        final FileLock lock;
+        final long owningThreadId;
+
+        LockAndOwner(long owningThreadId, FileLock lock) {
+            this.owningThreadId = owningThreadId;
+            this.lock = lock;
+        }
+    }
+
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: For line 65 and line 69: seems we can collapse them into one constructor as `threadId` is not useful anymore.

**Comment by guozhangwang** [comment]
Another meta question: could you try running `StreamThreadTest.shouldReleaseStateDirLockIfFailureOnTaskSuspend` locally and see if the previous issue we saw about `if (!locks.containsKey(id)) ` returns true even for the same thread which has already get the lock, has gone away?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 13]
```diff
@@ -95,20 +103,29 @@ File globalStateDir() {
...
      */
-    boolean lock(final TaskId taskId, int retry) throws IOException {
+    synchronized boolean lock(final TaskId taskId, int retry) throws IOException {
+
         final File lockFile;
         // we already have the lock so bail out here
-        if (locks.containsKey(taskId)) {
-            log.trace("{} Found cached state dir lock for task {}", logPrefix, taskId);
+        final LockAndOwner lockAndOwner = locks.get(taskId);
+        if (lockAndOwner != null && lockAndOwner.owningThreadId == Thread.currentThread().getId()) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: I'm wondering why use thread id instead of thread name here? Is the thread id from `getId()` guaranteed to be unique globally?

**Review by guozhangwang — COMMENTED** [review]
@dguy Could we run some simple benchmark with appropriate number (say, 8?) of threads and relatively small cleanup interval see if there is any unexpected perf differences?

Otherwise just minor comments.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java (hunk 13 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: we can use name. The id is unique during the lifetime of the thread but may be reused if it died. That wouldn't be a problem at the moment as we never start new threads, but might be better to use the name anyway

**Comment by dguy** [comment]
@guozhangwang w.r.t `StreamThreadTest.shouldReleaseStateDirLockIfFailureOnTaskSuspend` the test has changed to use a mock `StateDirectory` so this definitely won't occur in that test.

Also, the methods on `StateDirectory` are now synchronized so i don't see any possibility of this happening.

**Comment by dguy** [comment]
@guozhangwang below are results from running the simple benchmark with 8 threads. Looks like no difference really

**cleanup thread every 100ms**
Streams Performance [records/latency/rec-sec/MB-sec source+store]: 10000000/18070/553403.4311012728/69.42790970329956
Streams Performance [records/latency/rec-sec/MB-sec source+store]: 10000000/18197/549541.1331538166/69.00844394738134
Streams Performance [records/latency/rec-sec/MB-sec source+store]: 10000000/17967/556575.9447876663/69.85705100667957


**never cleanup**
Streams Performance [records/latency/rec-sec/MB-sec source+store]: 10000000/18150/550964.1873278237/69.1831415349787
Streams Performance [records/latency/rec-sec/MB-sec source+store]: 10000000/18455/541858.5749119481/67.98810294830066
Streams Performance [records/latency/rec-sec/MB-sec source+store]: 10000000/17897/558752.8636084261/70.0544206071794

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6309/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6293/
Test PASSed (JDK 8 and Scala 2.12).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java (hunk 13 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: Cool, let's use the name then.

**Review by guozhangwang — APPROVED** [review]
Besides thread-name v.s. thread-id in the locks map, otherwise LGTM. Please feel free to merge once that is addressed / PR is rebased.

**Comment by dguy** [comment]
retest this please

**Comment by dguy** [comment]
retest this please

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6343/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6328/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by dguy** [comment]
Merged to trunk

**Comment by guozhangwang** [comment]
@dguy I realize it is not cherry-picked to 0.11.0; could you file a separate PR for it?
